File: src/ui_launchers/Karen-AI-Theme/src/realtime/topics.ts
/**
 * Topic registry for frontend.
 * Centralizes topic names with deterministic formatting.
 * Prevents arbitrary browser-supplied prefixes.
 */



const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

function validateUuid(value: string, name: string): string {
  if (!UUID_RE.test(value)) {
    throw new Error(`${name} must be a valid UUID, got ${value}`);
  }
  return value;
}

export function userTopic(tenantId: string, userId: string): string {
  return `tenant:${validateUuid(tenantId, "tenantId")}:user:${validateUuid(userId, "userId")}`;
}

export function conversationTopic(tenantId: string, conversationId: string): string {
  return `tenant:${validateUuid(tenantId, "tenantId")}:conversation:${validateUuid(conversationId, "conversationId")}`;
}

export function executionTopic(tenantId: string, executionId: string): string {
  return `tenant:${validateUuid(tenantId, "tenantId")}:execution:${validateUuid(executionId, "executionId")}`;
}

export function tenantNotificationsTopic(tenantId: string): string {
  return `tenant:${validateUuid(tenantId, "tenantId")}:notifications`;
}

export function adminTopic(tenantId: string): string {
  return `tenant:${validateUuid(tenantId, "tenantId")}:admin`;
}

export function resolveTopic(topic: string): string | null {
  const parts = topic.split(":");
  if (parts.length < 3 || parts.length > 4 || parts[0] !== "tenant") return null;
  if (!UUID_RE.test(parts[1])) return null;
  const kind = parts[2];
  if (!["user", "conversation", "execution", "notifications", "admin"].includes(kind)) return null;
  if (["user", "conversation", "execution"].includes(kind) && parts.length !== 4) return null;
  if (["user", "conversation", "execution"].includes(kind) && !UUID_RE.test(parts[3])) return null;
  return topic;
}

export class TopicRegistry {
  constructor(private tenantId: string) {}

  user(userId: string) {
    return userTopic(this.tenantId, userId);
  }
  conversation(conversationId: string) {
    return conversationTopic(this.tenantId, conversationId);
  }
  execution(executionId: string) {
    return executionTopic(this.tenantId, executionId);
  }
  notifications() {
    return tenantNotificationsTopic(this.tenantId);
  }
  admin() {
    return adminTopic(this.tenantId);
  }
}

File: src/ai_karen_engine/core/realtime/topics.py
"""Realtime topic registry.

Centralizes Realtime topic names with deterministic formatting.
No arbitrary browser-supplied prefixes allowed.
"""

from __future__ import annotations

import re
import uuid
from dataclasses import dataclass, field
from typing import Optional


_TOPIC_ID_PATTERN = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", re.I)


def _validate_uuid(value: str, name: str) -> uuid.UUID:
    if not _TOPIC_ID_PATTERN.match(value):
        raise ValueError(f"{name} must be a valid UUID, got {value!r}")
    return uuid.UUID(value)


@dataclass(frozen=True)
class RealtimeTopicFactory:
    """Canonical topic name factory.

    Canonical topic format:
        tenant:{tenant_id}:user:{user_id}
        tenant:{tenant_id}:conversation:{conversation_id}
        tenant:{tenant_id}:execution:{execution_id}
        tenant:{tenant_id}:notifications
    """

    tenant_id: str
    _validated_tenant: uuid.UUID = field(init=False)

    def __post_init__(self) -> None:
        object.__setattr__(self, "_validated_tenant", _validate_uuid(self.tenant_id, "tenant_id"))

    def user_topic(self, user_id: str) -> str:
        uid = _validate_uuid(user_id, "user_id")
        return f"tenant:{self._validated_tenant}:user:{uid}"

    def conversation_topic(self, conversation_id: str) -> str:
        cid = _validate_uuid(conversation_id, "conversation_id")
        return f"tenant:{self._validated_tenant}:conversation:{cid}"

    def execution_topic(self, execution_id: str) -> str:
        eid = _validate_uuid(execution_id, "execution_id")
        return f"tenant:{self._validated_tenant}:execution:{eid}"

    def tenant_notifications_topic(self) -> str:
        return f"tenant:{self._validated_tenant}:notifications"

    def admin_topic(self) -> str:
        return f"tenant:{self._validated_tenant}:admin"

    def resolve_topic(self, topic: str) -> Optional[str]:
        """Validate and normalize a topic string.

        Returns the normalized topic if valid, None if invalid.
        """
        if not isinstance(topic, str):
            return None
        parts = topic.split(":")
        if len(parts) not in (3, 4):
            return None
        prefix = parts[0]
        if prefix != "tenant":
            return None
        if not _TOPIC_ID_PATTERN.match(parts[1]):
            return None
        kind = parts[2]
        valid_kinds = {"user", "conversation", "execution", "notifications", "admin"}
        if kind not in valid_kinds:
            return None
        if kind in {"user", "conversation", "execution"}:
            if len(parts) != 4 or not _TOPIC_ID_PATTERN.match(parts[3]):
                return None
        return topic
